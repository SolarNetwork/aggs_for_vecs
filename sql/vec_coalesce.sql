-- null scalar
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(ARRAY[1,2,3]::integer[], NULL)) r(r);
-- ints
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(ints, 0) FROM measurements WHERE sensor_id BETWEEN 1 AND 5) r(r);
-- floats
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(floats, 0) FROM measurements WHERE sensor_id BETWEEN 1 AND 5) r(r);
--numeric
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(nums, 0) FROM measurements WHERE sensor_id BETWEEN 1 AND 5) r(r);
