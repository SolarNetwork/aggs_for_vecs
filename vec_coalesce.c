
Datum vec_coalesce(PG_FUNCTION_ARGS);
PG_FUNCTION_INFO_V1(vec_coalesce);

/**
 * Coalesce elements in an array to a non-null value.
 *
 * This function takes an array of n elements plus a non-null value of the same type
 * as the array. It returns a new array with any NULL elements of the input array
 * replaced by the given non-null value.
 */
Datum
vec_coalesce(PG_FUNCTION_ARGS)
{
  Oid scalarTypeId = get_fn_expr_argtype(fcinfo->flinfo, 1);
  Oid elemTypeId;
  int16 elemTypeWidth;
  bool elemTypeByValue;
  char elemTypeAlignmentCode;
  int inputLength;
  ArrayType *inputArray, *retArray;
  Datum *inputContent, scalarContent, *retContent;
  bool *inputNulls, *retNulls;
  int i;
  int dims[1];
  int lbs[1];

  if (PG_ARGISNULL(0) || PG_ARGISNULL(1)) {
    PG_RETURN_NULL();
  }

  inputArray = PG_GETARG_ARRAYTYPE_P(0);

  if (ARR_NDIM(inputArray) == 0) {
    PG_RETURN_NULL();
  }
  if (ARR_NDIM(inputArray) > 1) {
    ereport(ERROR, (errmsg("vec_coalesce: one-dimensional array is required")));
  }

  elemTypeId = ARR_ELEMTYPE(inputArray);

  if (elemTypeId != INT2OID &&
      elemTypeId != INT4OID &&
      elemTypeId != INT8OID &&
      elemTypeId != FLOAT4OID &&
      elemTypeId != FLOAT8OID &&
      elemTypeId != NUMERICOID) {
    ereport(ERROR, (errmsg("vec_coalesce input must be array of SMALLINT, INTEGER, BIGINT, REAL, DOUBLE PRECISION, or NUMERIC")));
  }
  if (elemTypeId != scalarTypeId) {
    ereport(ERROR, (errmsg("vec_coalesce input array and scalar must be the same type")));
  }

  get_typlenbyvalalign(scalarTypeId, &elemTypeWidth, &elemTypeByValue, &elemTypeAlignmentCode);
  deconstruct_array(inputArray, scalarTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode,
      &inputContent, &inputNulls, &inputLength);

  retContent = palloc0(sizeof(Datum) * inputLength);
  retNulls = palloc0(sizeof(bool) * inputLength);

  scalarContent = PG_GETARG_DATUM(1);
  for (i = 0; i < inputLength; i++) {
    retNulls[i] = false;
    if (inputNulls[i]) {
      retContent[i] = scalarContent;
    } else {
      retContent[i] = inputContent[i];
    }
  }

  dims[0] = inputLength;
  lbs[0] = 1;

  retArray = construct_md_array(retContent, retNulls, 1, dims, lbs,
      scalarTypeId, elemTypeWidth, elemTypeByValue, elemTypeAlignmentCode);

  PG_RETURN_ARRAYTYPE_P(retArray);
}

