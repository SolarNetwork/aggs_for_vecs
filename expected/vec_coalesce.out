-- null scalar
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(ARRAY[1,2,3]::integer[], NULL)) r(r);
 r | pg_typeof 
---+-----------
   | integer[]
(1 row)

-- ints
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(ints, 0) FROM measurements WHERE sensor_id BETWEEN 1 AND 5) r(r);
    r    | pg_typeof 
---------+-----------
         | integer[]
 {0,0,0} | integer[]
 {1,2,3} | integer[]
 {1,0,2} | integer[]
 {1,2,4} | integer[]
(5 rows)

-- floats
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(floats, 0) FROM measurements WHERE sensor_id BETWEEN 1 AND 5) r(r);
    r    |     pg_typeof      
---------+--------------------
         | double precision[]
 {0,0,0} | double precision[]
 {1,2,3} | double precision[]
 {1,0,2} | double precision[]
 {1,2,4} | double precision[]
(5 rows)

--numeric
SELECT r, pg_typeof(r) FROM (SELECT vec_coalesce(nums, 0) FROM measurements WHERE sensor_id BETWEEN 1 AND 5) r(r);
        r         | pg_typeof 
------------------+-----------
                  | numeric[]
 {0,0,0}          | numeric[]
 {1.23,2.34,3.45} | numeric[]
 {1.23,0,2.34}    | numeric[]
 {1.23,2.34,4.45} | numeric[]
(5 rows)

